### PR TITLE
29 json support

### PR DESCRIPTION
**Pull request by Lorenzobattistela**
Implements #29 .

Implements a simple JSON parser. It correctly compiles to haskell through `agda-compile` command. 

The javascript compilation kinda works, but faces the same `bind` problem that is being investigated. (the js compilation is not  a problem since we can just use JSON.parse and JSON.stringify).
Example:

```agda
module Data.JSON.Main where

open import Data.String.Type
open import Data.String.show
open import Data.String.append
open import Data.JSON.Type
open import Data.JSON.parse
open import Data.Parser.State
open import Data.Parser.Error
open import Data.Parser.Reply
open import Data.Result.Type
open import Data.IO.print
open import Data.IO.Type
open import Data.IO.bind
open import Data.IO.seq
open import Data.Unit.Type
open import Data.Function.case
open import Data.JSON.show renaming (show to show-json)

main : IO Unit
main = do
  let json-string = "[1, true, null, [\"nested\", [2.5]], {\"key\": \"value\"}]"
  print ("Parsing JSON string: " ++ show json-string) 
  case parse-json-string json-string of λ where
    (Done reply) → do
      print "Parsing successful!"
      print ("Parsed JSON: " ++ show-json (Reply.value reply))
    (Fail error) → do
      print "Parsing failed!"
```

This compiles correctly to Haskell and parses the JSON. 

**NOTE**: the order of list elements is not preserved.

### CODE REVIEW

File: Data/JSON/parse-number.agda [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+parse-number : Parser JSON
+parse-number = do
+  skip-spaces
+  sign <- (consume "-" >> pure "-") <|> pure ""
+  int-part <- take-while is-digit
+  frac-part <- (do
+    consume "."
+    frac <- take-while is-digit
+    pure ("." ++ frac)) <|> pure ""
+
```
Comment (developedby) [inline Data/JSON/parse-number.agda:39]: Missing the exponent part.
Can be given as this redex `([eE][+-]?\d+)?`
See https://www.json.org/json-en.html

File: Data/JSON/parse-object.agda [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+open import Data.Parser.skip-spaces
+open import Data.Parser.alternative
+open import Data.Parser.parse-string renaming (parse-string to parse-pstring)
+open import Data.String.Type
+
+-- Parses a single key-value pair in a JSON object.
+parse-pair : Parser JSON → Parser (Pair String JSON)
+parse-pair parseJSON = do
+  skip-spaces
+  key ← parse-pstring
```
Comment (developedby) [inline Data/JSON/parse-object.agda:19]: This should parse a JSON string.

File: Data/JSON/parse-string.agda [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+open import Data.Parser.parse-string renaming (parse-string to parse-pstring)
+
+-- Parses a JSON string value.
+-- - Skips any leading whitespace.
+-- - Parses a quoted string.
+-- - Returns a JString JSON value.
+parse-string : Parser JSON
+parse-string = do
+  skip-spaces
+  str ← parse-pstring
```
Comment (developedby) [inline Data/JSON/parse-string.agda:17]: I don't know if we want to add support to it right now, but json strings accept some other things
https://www.json.org/json-en.html

File: Data/List/split-at-element.agda [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+open import Data.Pair.Type
+open import Data.Bool.Type
+open import Data.Function.case
+
+-- Splits a list at the first occurrence of the given element, removing that element.
+-- - eq: An equality function for the element type.
+-- - x: The element to split at.
+-- - xs: The input list.
+-- = A pair of lists: the part before the split element and the part after (excluding the split element).
+split-at-element : ∀ {A : Set} → (A → A → Bool) → A → List A → Pair (List A) (List A)
```
Comment (developedby) [inline Data/List/split-at-element.agda:13]: An arbitrry design choice, but perhaps a `(A -> Bool) -> List A -> Pair (List A) (List A)` would be better?

Checking every element for some condition. A bit more generic than this function, which would simply be `split-at-element eq x = split (eq x)`

**Review by developedby — CHANGES_REQUESTED** [review]

File: Data/JSON/parse-string.agda [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  d1 ← parse-hex-digit
+  d2 ← parse-hex-digit
+  d3 ← parse-hex-digit
+  d4 ← parse-hex-digit
+  pure (((d1 * 4096) + (d2 * 256)) + ((d3 * 16) + d4))
+
+-- Parses an escape sequence
+parse-escape : Parser Char
+parse-escape = do
+  consume "\\"
```
[the remark this answers is not in the document]
Comment (Lorenzobattistela) [reply Data/JSON/parse-string.agda:49]: Maybe we should make sure that the string follows this representation by running `show str` before parsing it, what do you think?

**Review by developedby — APPROVED** [review]

**Pull request merged**